crates/floresta-node/src/json_rpc/blockchain.rs

```diff
@@ -300,6 +321,80 @@ impl<Blockchain: RpcChain> RpcImpl<Blockchain> {
     // getmempoolinfo
     // getrawmempool
 
+    fn calculate_chainwork_by_height(&self, block_height: u32) -> Result<Work, JsonRpcError> {
+        let mut total_chainwork = Work::from_be_bytes([0u8; 32]);
+        for epoch_start_height in (0..=block_height).step_by(2016) {
+            // Calculate the number of blocks in this epoch
+            let epoch_end_height = min(epoch_start_height + 2015, block_height);
+            let blocks_in_epoch = (epoch_end_height - epoch_start_height + 1) as u64;
+
+            // Get the block hash and header at the start of the epoch
+            let epoch_block_hash = self
+                .chain
+                .get_block_hash(epoch_start_height)
+                .map_err(|_| JsonRpcError::BlockNotFound)?;
+            let epoch_block_header = self
+                .chain
+                .get_block_header(&epoch_block_hash)
+                .map_err(|_| JsonRpcError::BlockNotFound)?;
+
+            let epoch_chainwork =
+                Self::multiply_work_by_u64(epoch_block_header.work(), blocks_in_epoch)?;
+            total_chainwork = total_chainwork.add(epoch_chainwork);
+        }
+        Ok(total_chainwork)
+    }
+
+    fn multiply_work_by_u64(work: Work, factor: u64) -> Result<Work, JsonRpcError> {
+        if factor == 0 {
+            return Ok(Work::from_be_bytes([0u8; 32]));
+        }
+        if factor == 1 {
+            return Ok(work);
+        }
+
+        // Convert Work to little-endian bytes for easier manipulation (least significant byte first)
+        let work_bytes = work.to_le_bytes();
+        let mut carry_high: u128 = 0;
```

Comment:
I just thought about something: this won't work on 32 bit systems. I think we are supposed to support those.